src/backend/src/oisy_user.rs

```diff
@@ -0,0 +1,35 @@
+use crate::{Candid, StoredPrincipal, VMem};
+use candid::Principal;
+use ic_stable_structures::StableBTreeMap;
+use shared::types::{
+    user_profile::{GetUsersRequest, OisyUser, StoredUserProfile},
+    Timestamp,
+};
+use std::ops::Bound;
+
+const DEFAULT_LIMIT_GET_USERS_RESPONSE: usize = 10_000;
+
+pub fn get_oisy_users(
+    request: &GetUsersRequest,
+    user_profile_map: &StableBTreeMap<(u64, StoredPrincipal), Candid<StoredUserProfile>, VMem>,
+) -> (Vec<OisyUser>, u64) {
+    let limit_users_size: usize = match request.matches_max_length {
+        Some(num) => match usize::try_from(num) {
+            Ok(val) if val <= DEFAULT_LIMIT_GET_USERS_RESPONSE => val,
+            _ => DEFAULT_LIMIT_GET_USERS_RESPONSE,
+        },
+        None => DEFAULT_LIMIT_GET_USERS_RESPONSE,
+    };
+
+    let start_bound: Bound<(Timestamp, StoredPrincipal)> = match request.updated_after_timestamp {
+        Some(updated) => Bound::Included((updated, StoredPrincipal(Principal::anonymous()))),
+        None => Bound::Unbounded,
+    };
+    let users: Vec<OisyUser> = user_profile_map
+        .range((start_bound, Bound::Unbounded))
+        .take(limit_users_size)
```

Comment:
This isn't really performance, should we use a `limit_max_timestamp` instead of limiting the number of users that way we can limit the range?

Reply:
Why isn't this performant?

We discussed this with Frederik and Max in the document and offline.

If we use an expiration, we never know how many there will be and the size limit is important for two reasons:
* Size limit of the response. Messages have a limit of 2Mb.
* Instruction consumption limit. If the canister reads too many from stable memory, it might run on a limit of instructions. Therefore, we need to limit how much we read. Hence the `take`.

In the [document](https://docs.google.com/document/d/1R-VFUzHsXLewSqMxEpEH3kDWOGdPUQcwGP4Ia5ucB48/edit#heading=h.b9uawa2tmu48), we discussed filtering only the lower bound with a timestamp, and the upper bound with a number of elements.

Reply:
range + take + map is less performant than just range

but if you all three are fine with that, for sure so be it